File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/CLIService.java
package org.elastos.trinity.runtime;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;
import android.net.wifi.WifiManager;
import android.os.AsyncTask;
import android.os.Handler;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.ProtocolException;
import java.net.URL;
import java.util.UUID;

public class CLIService implements NsdManager.DiscoveryListener, NsdManager.ResolveListener {
    private static final String TAG = "CLIService";
    private static final String SERVICE_TYPE = "_trinitycli._tcp.";

    private static CLIService cliService = null;
    private Context context;
    private AppManager appManager;
    private NsdManager nsdManager;
    private Boolean shouldRestartSearching = true;
    private Boolean operationCompleted = false;
    private Boolean isStarted = false;

    CLIService() {
        this.appManager = AppManager.getShareInstance();
        this.context = appManager.activity;

        nsdManager = (NsdManager)context.getSystemService(Context.NSD_SERVICE);
    }


    static CLIService getShareInstance() {
        if (CLIService.cliService == null) {
            CLIService.cliService = new CLIService();
        }
        return CLIService.cliService;
    }

    public void start() {
        if (isStarted) {
            return;
        }
        isStarted = true;
        //TODO::
//        searchForServices();
    }

    public void  stop() {
        isStarted = false;
        //TODO::
//        stopSearching(false);
    }

    private void searchForServices() {
        new Handler().postDelayed(() -> {
                Log.d(TAG, "Searching for local CLI service...");

            WifiManager wifi = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
            WifiManager.MulticastLock multicastLock = wifi.createMulticastLock("multicastLock");
            multicastLock.setReferenceCounted(true);
            multicastLock.acquire();

                shouldRestartSearching = true;
                operationCompleted = false;
                nsdManager.discoverServices(SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, this);

                new Handler().postDelayed(() -> {
                    stopSearching(true);
                }, 20000);
            }, 5000);
    }

    private void stopSearching(Boolean shouldRestart) {
        Log.d(TAG, "Stopping service search.");

        shouldRestartSearching = shouldRestart;
        nsdManager.stopServiceDiscovery(this);
    }

    // Called as soon as service discovery begins.
    @Override
    public void onDiscoveryStarted(String regType) {
        Log.d(TAG, "Service discovery started");
    }

    @Override
    public void onServiceFound(NsdServiceInfo service) {
        Log.d(TAG, "Found CLI service on local network: "+service);

        nsdManager.resolveService(service, this);

        stopSearching(false);
    }

    @Override
    public void onServiceLost(NsdServiceInfo service) {
        // When the network service is no longer available.
        // Internal bookkeeping code goes here.
        Log.e(TAG, "service lost: " + service);
    }

    @Override
    public void onDiscoveryStopped(String serviceType) {
        Log.d(TAG, "Search stopped.");

        if (shouldRestartSearching) {
            // Restart searching
            searchForServices();
        }
    }

    @Override
    public void onStartDiscoveryFailed(String serviceType, int errorCode) {
        Log.e(TAG, "Discovery failed: Error code:" + errorCode);
        nsdManager.stopServiceDiscovery(this);
    }

    @Override
    public void onStopDiscoveryFailed(String serviceType, int errorCode) {
        Log.e(TAG, "Discovery failed: Error code:" + errorCode);
        nsdManager.stopServiceDiscovery(this);
    }

    @Override
    public void onResolveFailed(NsdServiceInfo nsdServiceInfo, int i) {
        Log.d(TAG, "Resolved failed");
    }

    @Override
    public void onServiceResolved(NsdServiceInfo nsdServiceInfo) {
        Log.d(TAG,"Service resolved");

        if (operationCompleted)
            return;

        // Got a resolved service info - we can call the service to get and install our EPK
        downloadEPK(nsdServiceInfo, (epkPath) -> {
            installEPK(epkPath);

            // Resume the bonjour search task for future EPKs.
            searchForServices();
        });
    }

    private interface DownloadListener {
        void onDownloadComplete(String epkPath);
    }

    @SuppressLint("StaticFieldLeak")
    private void downloadEPK(NsdServiceInfo nsdServiceInfo, DownloadListener completion) {
        String ipAddress = nsdServiceInfo.getHost().getHostAddress();
        if (ipAddress == null) {
            Log.d(TAG, "No IP address found for the service. Aborting EPK download.");
            return;
        }

        Log.d(TAG,"Downloading the remote EPK");

        // Watchguard to prevent downloading multiple times when IP address is resolved multiple times.
        operationCompleted = true;

        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... voids) {
                try {
                    String serviceEndpoint = "http://"+ipAddress+":"+nsdServiceInfo.getPort()+"/downloadepk";
                    URL url = new URL(serviceEndpoint);
                    HttpURLConnection c = (HttpURLConnection) url.openConnection();

                    c.setRequestMethod("GET");
                    c.setDoOutput(true);
                    c.connect();

                    if (c.getResponseCode() >= 200 && c.getResponseCode() < 400) {
                        Log.d(TAG, "EPK file downloaded successfully with status code: "+c.getResponseCode());

                        Log.d(TAG, "Requesting app manager to install the EPK");

                        String tempFileName = UUID.randomUUID().toString()+".epk";
                        String destPath = context.getCacheDir() + "/" + tempFileName;

                        File outputFile = new File(destPath);
                        FileOutputStream fos = new FileOutputStream(outputFile);
                        InputStream is = c.getInputStream();

                        byte[] buffer = new byte[4096];
                        int len;

                        while ((len = is.read(buffer)) != -1)
                        {
                            fos.write(buffer, 0, len);
                        }

                        fos.close();
                        is.close();

                        completion.onDownloadComplete(destPath);
                    }
                    else {
                        Log.d(TAG, "Failed to download EPK with HTTP error "+c.getResponseCode());
                    }
                } catch (Exception e) {
                    Log.d(TAG, "Exception while downloading the EPK file");
                    e.printStackTrace();
                }

                return null;
            }
        }.execute(null, null, null);
    }

    private void installEPK(String epkPath) {
        appManager.setInstallUri(epkPath, true);
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/CLIService.swift
//
//  CLIService.swift
//  elastOS
//
//  Created by Benjamin Piette on 28/02/2020.
//

import Foundation
import MultipeerConnectivity

public class CLIService: NSObject, NetServiceBrowserDelegate, NetServiceDelegate {
    private static var cliService: CLIService?;
    var serviceBrowser: NetServiceBrowser?
    var services = [NetService]()
    var shouldRestartSearching = true
    var operationCompleted = false
    let appManager: AppManager!
    var isStarted = false;
    
    override init() {
        self.appManager = AppManager.getShareInstance();
        super.init()
    }
    
    static func getShareInstance() -> CLIService {
        if (CLIService.cliService == nil) {
            CLIService.cliService = CLIService();
        }
        return CLIService.cliService!;
    }
    
    func start() {
        if (isStarted) {
            return;
        }
        isStarted = true;
        searchForServices()
    }
    
    func stop() {
        isStarted = false;
        stopSearching(shouldRestart: false);
    }
    
    private func log(_ str: String) {
        print("CLIService - \(str)")
    }
    
    private func searchForServices() {
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(5), execute: {
            self.log("Searching for local CLI service...")
            
            if (self.serviceBrowser == nil) {
                self.serviceBrowser = NetServiceBrowser()
                self.serviceBrowser!.delegate = self
            }
            
            self.shouldRestartSearching = true
            self.operationCompleted = false
            self.serviceBrowser!.searchForServices(ofType: "_trinitycli._tcp.", inDomain: "")
            
            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(20), execute: {
                self.stopSearching(shouldRestart: true)
            })
        })
    }
    
    private func stopSearching(shouldRestart: Bool) {
        guard serviceBrowser != nil else {
            return;
        }
        
        log("Stopping service search.")
        
        self.shouldRestartSearching = shouldRestart
        serviceBrowser!.stop()
    }
    
    public func netServiceBrowserDidStopSearch(_ browser: NetServiceBrowser) {
        log("Search stopped.")
        
        self.serviceBrowser = nil
        
        if self.shouldRestartSearching {
            // Restart searching
            self.searchForServices()
        }
    }
    
    public func netServiceBrowser(_ browser: NetServiceBrowser, didFind service: NetService, moreComing: Bool) {
        log("Found CLI service on local network: \(service)")
        
        services.append(service)
        service.delegate = self
        service.resolve(withTimeout:10)

        self.stopSearching(shouldRestart: false)
    }
    
    public func netServiceDidResolveAddress(_ service: NetService) {
        guard !operationCompleted else {
            return
        }
        
        // Got a resolved service info - we can call the service to get and install our EPK
        downloadEPK(usingService: service) { epkPath in
            self.log("Requesting app manager to install the EPK")
            self.installEPK(epkPath: epkPath)
            // Resume the bonjour search task for future EPKs.
            self.searchForServices()
        }
    }
    
    private func getServiceIPAddress(_ service: NetService) -> String? {
        guard service.addresses != nil else {
            return nil
        }
        
        for address in service.addresses! {
            var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
            address.withUnsafeBytes { ptr in
                guard let sockaddr_ptr = ptr.baseAddress?.assumingMemoryBound(to: sockaddr.self) else {
                    return
                }
                let sockaddr = sockaddr_ptr.pointee
                guard getnameinfo(sockaddr_ptr, socklen_t(sockaddr.sa_len), &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST) == 0 else {
                    return
                }
            }
            let ipAddress = String(cString:hostname)
            
            // Check if the resolved ip address is a ipv4 address, not ipv6.
            if ipAddress.contains(":") {
                continue
            }
            
            log("Resolved CLI service IP address: \(ipAddress) port: \(service.port)")
        
            return ipAddress
        }
        
        return nil
    }
    
    private func downloadEPK(usingService service: NetService, completion: @escaping (String)->Void) {
        // Compute the service's download EPK URL
        guard let ipAddress = getServiceIPAddress(service) else {
            log("No IP address found for the service. Aborting EPK download.")
            return
        }
        
        // Watchguard to prevent downloading multiple times when IP address is resolved multiple times.
        self.operationCompleted = true
        
        let serviceEndpoint = "http://\(ipAddress):\(service.port)/downloadepk"
        
        // Call the service to download the EPK file
        let sessionConfig = URLSessionConfiguration.default
        let session = URLSession(configuration: sessionConfig)
        let url = URL(string: serviceEndpoint)!
        let request = try! URLRequest(url: url, method: .get)

        log("Downloading EPK file at url \(serviceEndpoint)")
        let task = session.downloadTask(with: request) { (tempLocalUrl, response, error) in
            if let tempLocalUrl = tempLocalUrl, error == nil {
                // Success
                if let statusCode = (response as? HTTPURLResponse)?.statusCode {
                    if (statusCode >= 200 && statusCode < 400) {
                        self.log("EPK file downloaded successfully with status code: \(statusCode)")
                        
                        // Copy EPK file to a temp location as it's going to be deleted after download by the session.
                        let destTempUrl = FileManager.default.temporaryFileURL(fileName: "appinstall.epk")!
                        do {
                            // Delete the target file in case it already exists
                            try? FileManager.default.removeItem(at: destTempUrl)
                            // Copy the downloaded file to a file that won't be deleted by the network session
                            try FileManager.default.copyItem(at: tempLocalUrl, to: destTempUrl)
                            
                            completion(destTempUrl.absoluteString)
                        }
                        catch {
                            self.log("Failed copy the EPK file. \(error)")
                        }
                    }
                    else {
                        self.log("Failed to download EPK with HTTP error \(statusCode)")
                    }
                }
                else {
                    self.log("Failed to download EPK with unknown HTTP error")
                }
            } else {
                self.log("Failure: \(error?.localizedDescription ?? "?")");
            }
        }
        task.resume()
    }
     
    private func installEPK(epkPath: String) {
        DispatchQueue.main.async {
            do {
                _ = try AppManager.getShareInstance().install(epkPath, true)
            }
            catch AppError.error(let err) {
                alertDialog("Install Error", err);
            } catch let error {
                alertDialog("Install Error", error.localizedDescription);
            }
        }
    }
}
